### ISSUE TITLE
podman run uses a different image than what was just built by podman build

### ISSUE THREAD

**Issue by omajid** [containers/podman#12761]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

If I build a container using `podman build` and give it a name like "d", often running the container runs a different container than the just-built one. It seems like the name stays associated with some other container image?

This bug doesn't always happen for me, but when it does, it's repeatable: I can repeat the same build commands and run commands and the wrong container is consistently used for the same build/run commands.
 
Cleaning all images generally fixes the issue, until it re-appears a few days/weeks later.

**Steps to reproduce the issue:**

1. Create a `Dockerfile`. This one is using `FROM quay.io/centos/centos:stream9-development`.
2. Build it:  `podman build -t d -f Dockerfile`
2. Run the image and show the OS version: `podman run -it d /usr/bin/cat /etc/os-release`. Notice it prints the wrong OS version. It's not CentOS Stream 9 like I expected based on the `FROM` line.

**Describe the results you received:**

```
$ podman build -t d -f src/centos/stream9/Dockerfile
STEP 1/2: FROM quay.io/centos/centos:stream9-development
STEP 2/2: RUN dnf install --setopt tsflags=nodocs --refresh -y         dnf-plugins-core     &&     dnf config-manager --set-enabled crb     &&     dnf install --setopt tsflags=nodocs -y         "perl(Time::HiRes)"         autoconf         automake         clang         cmake         curl-devel         doxygen         findutils         gcc         gdb         git         glibc-langpack-en         hostname         krb5-devel         libcurl-devel         libedit-devel         libicu-devel         libidn2-devel         libnghttp2-devel         libtool         libuuid-devel         libxml2-devel         lld         lldb-devel         llvm         lttng-ust-devel         lzma         make         ncurses-devel         numactl-devel         openssl-devel         procps-ng         python3         python3-devel         readline-devel         sudo         swig         tar         wget         which         xz         zlib-devel     &&     dnf clean all
--> Using cache e8806dd63c4d34c881d03b7a4fad0690e2390fe0255bf2948855252cbf38fd29
COMMIT d
--> e8806dd63c4
Successfully tagged localhost/d:latest
e8806dd63c4d34c881d03b7a4fad0690e2390fe0255bf2948855252cbf38fd29
$ podman run -it d /usr/bin/cat /etc/os-release
NAME="Fedora Linux"
VERSION="36 (Container Image Prerelease)"
ID=fedora
VERSION_ID=36
VERSION_CODENAME=""
PLATFORM_ID="platform:f36"
PRETTY_NAME="Fedora Linux 36 (Container Image Prerelease)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:36"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/rawhide/system-administrators-guide/"
SUPPORT_URL="https://ask.fedoraproject.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=rawhide
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=rawhide
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Container Image"
VARIANT_ID=container
$ podman images --all
REPOSITORY                         TAG                  IMAGE ID      CREATED            SIZE
localhost/d                        latest               e8806dd63c4d  About an hour ago  1.36 GB
quay.io/centos/centos              stream9-development  a0139f353193  14 hours ago       146 MB
quay.io/centos/centos              stream9              b575d3c08f6e  2 weeks ago        152 MB
registry.access.redhat.com/ubi8    latest               fca12da1dc30  3 weeks ago        235 MB
registry.fedoraproject.org/fedora  rawhide              dfe12cab192e  3 weeks ago        166 MB
registry.fedoraproject.org/fedora  35                   3059bef432eb  6 weeks ago        159 MB
registry.fedoraproject.org/fedora  34                   b0e38cd3ecc2  6 weeks ago        184 MB
quay.io/centos/centos              8                    5d0da3dc9764  3 months ago       239 MB
registry.centos.org/centos         8                    e64b6e20a604  13 months ago      217 MB

```

**Describe the results you expected:**

`podman run` would run the just-built container image, not some other image.

**Additional information you deem important (e.g. issue happens only occasionally):**

- I don't have a perfect way to reproduce this, this only happens sometimes. But once it does, repeating the build and run commands doesn't fix it.
- Using a `localhost/` prefix  (like `localhost/d` in the example) seems to use the right container image.

**Output of `podman version`:**

```
$ podman version
Version:      3.4.4
API Version:  3.4.4
Go Version:   go1.16.8
Built:        Wed Dec  8 16:45:07 2021
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
$ podman info --debug                                                                     [43/2120]
host:                   
  arch: amd64          
  buildahVersion: 1.23.1                                                       
  cgroupControllers:   
  - memory                         
  - pids               
  cgroupManager: systemd                                                       
  cgroupVersion: v2                                                            
  conmon:                                                                      
    package: conmon-2.0.30-2.fc35.x86_64                                       
    path: /usr/bin/conmon                                                      
    version: 'conmon version 2.0.30, commit: '                                 
  cpus: 8               
  distribution:                                                                
    distribution: fedora                                                                                                                                      
    variant: workstation  
    version: "35"                                                                                                                                             
  eventLogger: journald                                                        
  hostname: terminus    
  idMappings:                                                                  
    gidmap:               
    - container_id: 0               
      host_id: 1000                                                            
      size: 1                       
    - container_id: 1                                                          
      host_id: 100000                                                          
      size: 65536               
    uidmap:                                                                    
    - container_id: 0  
      host_id: 1000              
      size: 1                       
    - container_id: 1                                                          
      host_id: 100000            
      size: 65536                                                              
  kernel: 5.15.11-200.fc35.x86_64
  linkmode: dynamic                  
  logDriver: journald
  memFree: 2627309568
  memTotal: 33391194112
  ociRuntime:
    name: crun
    package: crun-1.3-1.fc35.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.3
      commit: 8e5757a4e68590326dafe8a8b1b4a584b10a1370
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.12-2.fc35.x86_64
    version: |-
      slirp4netns version 1.1.12
      commit: 7a104a101aa3278a2152351a082a6df71f57c9a3
      libslirp: 4.6.1
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.3
  swapFree: 8431071232
  swapTotal: 8589930496
  uptime: 50h 54m 28.46s (Approximately 2.08 days)
plugins:               
  log:                           
  - k8s-file                        
  - none                                                                       
  - journald                     
  network:                                                                     
  - bridge                       
  - macvlan                          
  volume:            
  - local            
registries:            
  search:    
  - registry.fedoraproject.org
  - registry.access.redhat.com     
  - docker.io          
  - quay.io    
store:                
  configFile: /home/omajid/.config/containers/storage.conf                     
  containerStore:             
    number: 29                                                                 
    paused: 0
    running: 0 
    stopped: 29                                                                
  graphDriverName: overlay                                                     
  graphOptions: {}        
  graphRoot: /home/omajid/.local/share/containers/storage                                                                                                     
  graphStatus:    
    Backing Filesystem: btrfs
    Native Overlay Diff: "true"                                                
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:                                                                  
    number: 9                       
  runRoot: /run/user/1000/containers                                           
  volumePath: /home/omajid/.local/share/containers/storage/volumes
version:                        
  APIVersion: 3.4.4                                                            
  Built: 1638999907  
  BuiltTime: Wed Dec  8 16:45:07 2021
  GitCommit: ""                     
  GoVersion: go1.16.8                                                          
  OsArch: linux/amd64  
  Version: 3.4.4                                                               
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
$ rpm -q podman 
podman-3.4.4-1.fc35.x86_64

```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

A physical x86_64 machine running Fedora 35. Problem also existed in older Fedora versions (and older podman versions, I assume).

**Comment by vrothberg** [comment]
Thank you for reaching out!  Are there other services running on your machine which may be using the same image `d`?

The reported randomness in reproducibility suggests a race and I am fairly sure there is none in the code.

**Comment by omajid** [comment]
I am running into this again today and I don't see any duplicate images. Today, I am using `f`:

```
$ podman build -t f -f dotnet6.0-fedora.dockerfile
STEP 1/2: FROM registry.fedoraproject.org/fedora:35
STEP 2/2: RUN dnf install -yq       clang       cmake       coreutils       findutils       git       glibc-langpack-en       hostname       krb5-devel       libicu-devel       libunwind-devel       lld       llvm       lttng-ust-devel       make       openssl-devel       python3       tar       util-linux       zlib-devel
--> Using cache a6a00f7191e0e1daa8480f6011c5f2c13985a8108cc48c45f2942f0f3db2ef1a
COMMIT f
--> a6a00f7191e
Successfully tagged localhost/f:latest
a6a00f7191e0e1daa8480f6011c5f2c13985a8108cc48c45f2942
$ podman run -it f /bin/bash
[root@4d7352ec1965 /]# git
bash: git: command not found
[root@4d7352ec1965 /]# exit
```

I built a Fedora container, including `git` and `git` isn't found here. Looks like it's using another container instead:

```
$ podman ps -a | grep 4d7352ec1965
4d7352ec1965  registry.access.redhat.com/ubi8:latest     /bin/bash             About a minute ago  Exited (127) About a minute ago              heuristic_rhodes
```

I don't have any other containers running, at least that I can see via `podman ps`:

```
$ podman ps
CONTAINER ID  IMAGE       COMMAND     CREATED     STATUS      PORTS       NAMES
$ podman images | grep 'f'
localhost/f                        latest               a6a00f7191e0  17 minutes ago  1.11 GB
localhost/a                        latest               af996b398a90  7 hours ago     563 MB
<none>                             <none>               019e65ff9940  7 hours ago     563 MB
<none>                             <none>               a372f153b124  4 days ago      703 MB
<none>                             <none>               8bc299f09457  4 days ago      702 MB
<none>                             <none>               0ce80fa66c94  4 days ago      702 MB
quay.io/centos/centos              stream9-development  a0139f353193  12 days ago     146 MB
quay.io/centos/centos              stream9              b575d3c08f6e  4 weeks ago     152 MB
registry.access.redhat.com/ubi8    latest               fca12da1dc30  5 weeks ago     235 MB
registry.fedoraproject.org/fedora  rawhide              dfe12cab192e  5 weeks ago     166 MB
registry.fedoraproject.org/fedora  36                   dfe12cab192e  5 weeks ago     166 MB
registry.fedoraproject.org/fedora  35                   3059bef432eb  7 weeks ago     159 MB
registry.fedoraproject.org/fedora  34                   b0e38cd3ecc2  7 weeks ago     184 MB
docker.io/library/alpine           latest               c059bfaa849c  7 weeks ago     5.87 MB
```

**Comment by rhatdan** [comment]
Could you inspect your latest container and make sure it is built on  image a6a00f7191e0

**Comment by rhatdan** [comment]
You could also
$ podman unshare
# mnt=$(podman image mount f)
# ls $/mnt/bin/git

**Comment by omajid** [comment]
> Could you inspect your latest container and make sure it is built on image a6a00f7191e0

What's the right way to do this?

> You could also ...

I have a few other workarounds too. I would really like to understand the bug, though (and hopefully get it resolved too).

**Comment by afro-coder** [comment]
What's the right way to do this?
> podman inspect image-id

Check the ID part there

**Comment by omajid** [comment]
```
$ podman inspect f
Error: more than one result for container ID f: container already exists
```

**Comment by afro-coder** [comment]
That seems a bit weird can you try
```
podman image inspect name/id
```
Do you have any container with the same name? Try the ID instead

```
podman container ls -a
```

**Comment by omajid** [comment]
`podman inspect f` is working today, so I don't know if I can provide anything useful. 

What data should I collect when I next run into this error?

**Comment by rhatdan** [comment]
@vrothberg This looks like an error in containers/common/libimage.

**Comment by rhatdan** [comment]
We should default this to look at localhost/d

It looks like the name expansion of images is somehow effecting this.

I got this to fail and see this behavior.
```
$ podman run localhost/d cat /etc/redhat-release 
CentOS Stream release 9
$ podman run d cat /etc/redhat-release 
Fedora release 35 (Thirty Five)
 $ podman images
REPOSITORY                                TAG                  IMAGE ID      CREATED       SIZE
localhost/d                               latest               fe057e91ed62  17 hours ago  146 MB
quay.io/centos/centos                     stream9-development  fe057e91ed62  17 hours ago  146 MB
quay.io/podman/stable                     latest               d257063ef809  2 days ago    360 MB
registry.fedoraproject.org/fedora         latest               e417cd49a84e  12 days ago   159 MB
registry.access.redhat.com/ubi8/httpd-24  latest               028796994dbd  3 weeks ago   470 MB
registry.access.redhat.com/ubi8           latest               b81e86a2cb9a  3 weeks ago   235 MB
docker.io/library/alpine                  latest               c059bfaa849c  2 months ago  5.87 MB
quay.io/libpod/testimage                  20210610             9f9ec7f2fdef  8 months ago  7.99 MB
```

**Comment by rhatdan** [comment]
So the d is grabbing the fedora image which has IMAGE ID "d257063ef809"

**Comment by vrothberg** [comment]
As long as there is no clean reproducer, there is nothing I can do.

If possible, could you rerun with `podman --log-level=debug`? This will give use the logs for looking up the image in the local storage.

**Comment by omajid** [comment]
Thanks!

**Comment by vrothberg** [comment]
https://github.com/containers/common/pull/939 merged. The fix will make it into Podman 4.1.
